src/main.rs

```diff
@@ -7,7 +19,182 @@ use ic_cdk::api::management_canister::http_request::{HttpHeader, HttpResponse, T
+        METADATA.with(|m| m.borrow().get().nodes_in_subnet)
+    }
+
+    fn resolve_api(provider: RpcNodeProvider) -> Result<RpcApi, ProviderError> {
+        // fn resolve_by_hostname(hostname: &str) -> Result<RpcApi, ProviderError> {
+        //     unimplemented!()
+        // }
+        // use RpcNodeProvider::*;
+        // match provider {
+        //     // TODO
+        //     // Ethereum(EthereumProvider::PublicNode) => resolve_by_hostname(""),
+        // }
+        Ok(provider.api())
+    }
+
+    async fn call_json_rpc<T: DeserializeOwned>(
+        provider: RpcNodeProvider,
+        json: &str,
+        max_response_bytes: u64,
+    ) -> Result<T, RpcError> {
+        let response = do_http_request(
+            ic_cdk::caller(),
+            ResolvedSource::Api(Self::resolve_api(provider)?),
+            json,
+            max_response_bytes,
+        )
+        .await
+        .unwrap();
+        let status = get_http_response_status(response.status.clone());
+        let body = get_http_response_body(response)?;
+        let json: JsonRpcReply<T> = serde_json::from_str(&body).unwrap_or_else(|e| {
+            Err(HttpOutcallError::InvalidHttpJsonRpcResponse {
+                status,
+                body,
+                parsing_error: Some(format!("JSON response parse error: {e}")),
+            })
+        })?;
+        json.result.into()
+    }
+}
+
+fn get_rpc_client(source: CandidRpcSource) -> Result<EthRpcClient<CanisterTransport>, RpcError> {
+    fn validate_providers<T>(opt_vec: Option<Vec<T>>) -> Result<Option<Vec<T>>, RpcError> {
+        Ok(match opt_vec {
+            Some(v) if v.is_empty() => Err(ProviderError::ProviderNotFound)?,
+            opt => opt,
+        })
+    }
+    if !is_rpc_allowed(&ic_cdk::caller()) {
+        // inc_metric!(eth_*_err_no_permission);
+        return Err(ProviderError::NoPermission.into());
+    }
+    Ok(match source {
+        CandidRpcSource::EthMainnet(service) => EthRpcClient::new(
+            EthereumNetwork::Ethereum,
+            validate_providers(Some(vec![service.unwrap_or(
+                cketh_common::eth_rpc_client::providers::EthereumProvider::Ankr,
+            )]))?
+            .map(|p| p.into_iter().map(RpcNodeProvider::Ethereum).collect()),
+        ),
+        CandidRpcSource::EthSepolia(service) => EthRpcClient::new(
+            EthereumNetwork::Sepolia,
+            validate_providers(Some(vec![service.unwrap_or(
+                cketh_common::eth_rpc_client::providers::SepoliaProvider::PublicNode,
+            )]))?
+            .map(|p| p.into_iter().map(RpcNodeProvider::Sepolia).collect()),
+        ),
+    })
+}
+
+fn wrap_result<T>(result: Result<T, MultiCallError<T>>) -> RpcResult<T> {
+    match result {
+        Ok(value) => Ok(value),
+        Err(err) => match err {
+            MultiCallError::ConsistentError(e) => Err(e),
+            MultiCallError::InconsistentResults(_r) => {
+                unreachable!()
```

Comment:
Why is it unreachable?

Reply:
This is an artifact from the structure of the ckETH minter codebase, which uses multiple results. Since we recently removed this feature from the scope of the EVM RPC canister, we now perform a multi-call result with just one RPC provider, so there's no way to get inconsistent results. 

In other words, this logic made more sense when we called multiple providers in the Candid-RPC methods and now requires some additional refactoring.